### PR TITLE
List hardware vs associated.

### PR DESCRIPTION
**Pull request by FernandoOjeda**
List hardware vs associated https://github.com/softlayer/softlayer-python/issues/1301.
Add list hardware vs associated
`slcli hardware guests
`
Add vs list hw vs associated.
`slcli vs list`

### CODE REVIEW

File: SoftLayer/CLI/hardware/guests.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+from SoftLayer.CLI import formatting
+from SoftLayer.CLI import helpers
+from SoftLayer import utils
+
+
+@click.command()
+@click.argument('identifier')
+@environment.pass_env
+def cli(env, identifier):
+    """List the Hardware server associated virtual guests."""
```
Comment (allmightyspiff) [inline SoftLayer/CLI/hardware/guests.py:18]: Change  this, and the top line to

> Lists the Virtual Guests running on this server.

File: SoftLayer/CLI/hardware/guests.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    mgr = SoftLayer.HardwareManager(env.client)
+    hw_id = helpers.resolve_id(mgr.resolve_ids, identifier, 'hardware')
+    hw_guests = mgr.get_hardware_guests(hw_id)
+
+    if not hw_guests:
+        raise exceptions.CLIAbort("The hardware server does not has associated virtual guests.")
```
Comment (allmightyspiff) [inline SoftLayer/CLI/hardware/guests.py:25]: Change to 

> No Virtual Guests found.

File: SoftLayer/managers/vs.py [hunk 3]
```diff
@@ -1158,3 +1158,12 @@ def get_local_disks(self, instance_id):
         """
         mask = 'mask[diskImage]'
         return self.guest.getBlockDevices(mask=mask, id=instance_id)
+
+    def get_hardware_guests(self):
+        """Returns the hardware server vs associated.
+
+        :return SoftLayer_Hardware[].
+        """
+        object_filter = {"hardware": {"networkGatewayMemberFlag": {"operation": 0}}}
```
Comment (allmightyspiff) [inline SoftLayer/managers/vs.py:1167]: `'{"hardware":{"virtualHost":{"id":{"operation":"not null"}}}}`  might be a better filter.

File: SoftLayer/managers/vs.py [hunk 4]
```diff
@@ -1158,3 +1158,12 @@ def get_local_disks(self, instance_id):
         """
         mask = 'mask[diskImage]'
         return self.guest.getBlockDevices(mask=mask, id=instance_id)
+
+    def get_hardware_guests(self):
+        """Returns the hardware server vs associated.
```
Comment (allmightyspiff) [inline SoftLayer/managers/vs.py:1163]: > Returns all virtualHost capable hardware objects and their guests.

File: tests/managers/hardware_tests.py [hunk 5]
```diff
@@ -626,6 +626,54 @@ def test_get_hard_drive_empty(self):
...
+                },
+                "status": {
+                    "keyName": "ACTIVE",
+                    "name": "Active"
+                }
+            }]
+
+        result = self.hardware.get_hardware_guests(1234)
+
+        self.assertEqual([
```
Comment (allmightyspiff) [inline tests/managers/hardware_tests.py:659]: I'd prefer tests like this only test a few of the expected values, not testing if the whole object is an expected value. That way adding more data won't break the test.

`self.assertEqual("NSX-T Manager", result[0]['hostname'])`

or something like that.

File: tests/managers/vs/vs_tests.py [hunk 6]
```diff
@@ -1132,3 +1132,61 @@ def test_get_local_disks_swap(self):
...
+                        },
+                        "status": {
+                            "keyName": "ACTIVE",
+                            "name": "Active"
+                        }
+                    }]}}]
+
+        result = self.vs.get_hardware_guests()
+
+        self.assertEqual([{
```
Comment (allmightyspiff) [inline tests/managers/vs/vs_tests.py:1167]: Use a more precise assertion here. Just test a few fields that you expect to exist, don't need to test the entire object is strictly equal.

File: SoftLayer/CLI/virt/list.py [hunk 7]
```diff
@@ -93,3 +93,23 @@ def cli(env, sortby, cpu, domain, datacenter, hostname, memory, network,
                        for value in columns.row(guest)])
 
     env.fout(table)
+
+    hardware_guests = vsi.get_hardware_guests()
+    for hardware in hardware_guests:
+        if 'virtualHost' in hardware and hardware['virtualHost']['guests']:
+            table_hardware_guest = formatting.Table(['id', 'hostname', 'CPU', 'Memory', 'Start Date', 'Status',
+                                                     'powerState'], title="Hardware(id = {hardwareId}) guests "
```
Comment (allmightyspiff) [inline SoftLayer/CLI/virt/list.py:101]: to make this line a little easier to read, set the title to a variable and just use that here.

File: SoftLayer/CLI/hardware/guests.py (hunk 2 above)
Comment (FernandoOjeda) [reply SoftLayer/CLI/hardware/guests.py]: Done.

File: SoftLayer/managers/vs.py (hunk 3 above)
Comment (FernandoOjeda) [reply SoftLayer/managers/vs.py]: Done.

File: SoftLayer/managers/vs.py (hunk 4 above)
Comment (FernandoOjeda) [reply SoftLayer/managers/vs.py]: Done.

File: tests/managers/hardware_tests.py (hunk 5 above)
Comment (FernandoOjeda) [reply tests/managers/hardware_tests.py]: Done.

File: tests/managers/vs/vs_tests.py (hunk 6 above)
Comment (FernandoOjeda) [reply tests/managers/vs/vs_tests.py]: Done.

File: SoftLayer/CLI/virt/list.py (hunk 7 above)
Comment (FernandoOjeda) [reply SoftLayer/CLI/virt/list.py]: Done.

**Pull request merged**